devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+# `waiter` requirements
+
+## Overview
+
+`waiter` is a module that allows the user to move data asynchronously from producer to consumer without having to block the thread.
+
+Consumers of data call `waiter_register_notification` to register a `NOTIFICATION_CALLBACK` function to be called when data becomes available.
+
+Producers of data call `waiter_notify` to notify the waiter that data is available.
+
+`waiter_register_notification` and `waiter_notify` can be called from different threads in any order.
+
+## `waiter` state
+
+The state of the `waiter` module consists of three components:
+- `EPOCH`: This is a monotonically increasing integer that is incremented every time an operation is completed or cancelled.
+- `NOTIFIED`: This is a bit signifying that `waiter_notify` has been called previously and a `waiter_register_notification` call is expected to complete the operation.
+- `REGISTERED`: This is a bit signifying that `waiter_register_notification` has been called previously and a `waiter_notify` call is expected to complete the operation.
+
+All three components are stored in a single `volatile_atomic` variable that is manipulated using `interlocked` APIs.
+
+```mermaid
+stateDiagram-v2
+    B: EPOCH=x,N=0,R=0
+    N: EPOCH=x,N=1,R=0
+    R: EPOCH=x,N=0,R=1
+    E: EPOCH=x+1,N=0,R=0
+
+    B --> N:  waiter_notify
+    B --> R:  waiter_register_notification
+
+    N --> E:  waiter_register_notification
+    R --> E: waiter_notify
+
+    N --> E:  cancel waiter_notify
+    R --> E: cancel waiter_register_notification
+
+    R --> R: waiter_register_notification
+    N --> N: waiter_notify
+```
+
+## Reentrancy
+
+`waiter` is reentrant. This means that `waiter_register_notification` and `waiter_notify` can be called from `NOTIFICATION_CALLBACK` and `NOTIFY_COMPLETE_CALLBACK` functions.
```

Comment:
>can be called from [](http://example.com/codeflow?start=90&length=18)

I suppose that any API can be called from any callback, so if somehow the threads gets switched and on the producer's callback the consumer API is called and on the consumer's callback the producer API is called - they still work.

Reply:
Fixed.